### DISCUSSION TITLE
JWE Authentication

### DISCUSSION THREAD

**Discussion by frengooo** [mia-platform/community#100] [Q&A] [answered]
Hi guys,
a client configured the Envoy JWT authentication filter to authenticate requests through a JWT and Auth0, and it works fine.

However, for some request, the authentication is different, because uses the following flow, involving the JWE (Json Web Encryption):
- incoming request with a **JWE** instead of JWT
- the _payload_ of the JWE contains a field with the JWT
- the JWT must be validated with Auth0

Is possible to customize the Envoy HTTP filter, adding some kind of `pre request script` on it, to extract the JWT from the payload of the JWE? Any idea?

**Comment by davidebianchi** [comment]
Hi @silversoul93, maybe it's possible with a [lua script](https://docs.mia-platform.eu/docs/development_suite/api-console/advanced-section/api-gateway-envoy/extensions#lua-scripts), you can try to add some custom function.

**Comment by fiscafusca** [comment, marked as answer]
As @davidebianchi already suggested, you could try to implement a Lua script for the functionality you need.
Just remember that the sorting in the filter chain is critical and may change the behavior of the API Gateway. If you want the Lua script to perform manipulation on the JWE *before* the JWT authentication filter, you will probably need to explicitly set a higher priority to the Lua filter (check [this section](https://docs.mia-platform.eu/docs/development_suite/api-console/advanced-section/api-gateway-envoy/extensions#http-filters) for more info on HTTP filter ordering).

**Reply by frengooo** [reply]
Thanks @davidebianchi and @fiscafusca.
This flow should be perfect for this use case, we just need to decode the JWE to get the payload; do you know if there is a way to do this using Lua?
The used algorithm is A256GCM (_Advanced Encryption Standard (AES) in Galois/Counter Mode (GCM) algorithm with a 256-bit long key, and it’s a symmetric key algorithm used for AEAD_).

Moreover, it's possible to import external libraries in Lua script used with the Envoy api-gateway?

**Reply by fiscafusca** [reply]
Yes, you can! Do you want to import an existing library or a custom library?
If you want to add a custom library, you can create a ConfigMap containing the Lua library and mount it to the API Gateway deployment. 
Then, you can import the library in your code as follows:
```
local lib = require('/lua/path/to/lib')
```
And call the lib functions in the script like this:
```
lib.my_function()
```

**Reply by frengooo** [reply]
Thanks, we will look for libraries to decode a JWE in Lua 😄
